File: app/lib/features/documents/domain/uploaded_document.dart
class ExtractedField {
  const ExtractedField({required this.value, required this.confidence});

  final String value;
  final double confidence;

  factory ExtractedField.fromJson(Map<String, dynamic> json) =>
      ExtractedField(value: json['value'] as String, confidence: (json['confidence'] as num).toDouble());
}

class UploadedDocument {
  const UploadedDocument({
    required this.id,
    required this.fileName,
    required this.documentType,
    required this.typeConfidence,
    required this.status,
    required this.fields,
    this.errorMessage,
    this.mismatchWarning,
  });

  final String id;
  final String fileName;
  final String documentType;
  final double typeConfidence;

  /// "Pending" | "Processed" | "Failed"
  final String status;
  final String? errorMessage;

  /// Non-null when this document appears to be about a different
  /// real-world subject than what the user already told the system (e.g.
  /// a different car). The document and its fields are still saved and
  /// editable - they're just not silently used to fill the interview.
  final String? mismatchWarning;
  final Map<String, ExtractedField> fields;

  bool get isProcessed => status == 'Processed';
  bool get isFailed => status == 'Failed';

  /// Returns a copy with [key] set to a user-confirmed value (confidence
  /// 1.0 - a human just looked at the document and typed it themselves).
  UploadedDocument withField(String key, String value) {
    final updated = Map<String, ExtractedField>.from(fields);
    updated[key] = ExtractedField(value: value, confidence: 1);
    return UploadedDocument(
      id: id,
      fileName: fileName,
      documentType: documentType,
      typeConfidence: typeConfidence,
      status: status,
      errorMessage: errorMessage,
      mismatchWarning: mismatchWarning,
      fields: updated,
    );
  }

  factory UploadedDocument.fromJson(Map<String, dynamic> json) => UploadedDocument(
    id: json['id'] as String,
    fileName: json['fileName'] as String,
    documentType: json['documentType'] as String,
    typeConfidence: (json['typeConfidence'] as num).toDouble(),
    status: json['status'] as String,
    errorMessage: json['errorMessage'] as String?,
    mismatchWarning: json['mismatchWarning'] as String?,
    fields: (json['fields'] as Map<String, dynamic>).map(
      (key, value) => MapEntry(key, ExtractedField.fromJson(value as Map<String, dynamic>)),
    ),
  );
}

File: backend/src/EasyAgree.Domain/Entities/UploadedDocument.cs
using EasyAgree.Domain.Enums;

namespace EasyAgree.Domain.Entities;

/// <summary>
/// One file uploaded against a <see cref="Deal"/> (a passport photo, a
/// cadastre scan, etc). Classification and field extraction both happen
/// from a single vision-model pass over the file and are persisted here so
/// the interview planner can treat them the same way it treats the
/// original free-form request - already-known information to never ask
/// about again.
/// </summary>
public class UploadedDocument
{
    public Guid Id { get; set; }

    public Guid DealId { get; set; }

    public required string FileName { get; set; }

    public required string ContentType { get; set; }

    /// <summary>Path on disk (relative to the uploads root), not a public URL.</summary>
    public required string StoragePath { get; set; }

    public DocumentType DocumentType { get; set; } = DocumentType.Unknown;

    public double TypeConfidence { get; set; }

    /// <summary>
    /// Semantic key/value/confidence triples read from the document
    /// (e.g. <c>full_name</c>, <c>passport_number</c>, <c>vin</c>),
    /// serialized as JSON. Field keys are free-form on purpose - matching
    /// them to a specific template's field ids is left to the interview
    /// planner's existing natural-language extraction, the same mechanism
    /// already used for the original free-form request.
    /// </summary>
    public string? ExtractedFieldsJson { get; set; }

    /// <summary>
    /// Effective fields created after raw extraction (manual corrections,
    /// deterministic mapping and validation). This is deliberately separate
    /// from <see cref="ExtractedFieldsJson"/> so extraction can be replayed
    /// when mapping rules evolve without calling OCR/Vision again.
    /// </summary>
    public string? NormalizedFieldsJson { get; set; }

    /// <summary>Raw text read off the document - kept for transparency/debugging, not itself fed to the interview.</summary>
    public string? OcrText { get; set; }

    public DocumentProcessingStatus Status { get; set; } = DocumentProcessingStatus.Pending;

    public string? ErrorMessage { get; set; }

    /// <summary>
    /// Set when this document appears to be about a different real-world
    /// subject than what the user already told the system (e.g. a
    /// different car brand) - null otherwise. Purely informational: the
    /// document is still saved and its fields still show up in the
    /// editable-fields sheet, they're just not silently used to fill the
    /// interview.
    /// </summary>
    public string? MismatchWarning { get; set; }

    public DateTime UploadedAt { get; set; }

    public DateTime? ProcessedAt { get; set; }
}
